operator/internal/webhook/admission/clustertopology/validation/clustertopology.go

```diff
@@ -0,0 +1,188 @@
+	// (+kubebuilder:validation:Required, +kubebuilder:validation:MinLength=1, +kubebuilder:validation:MaxLength=63)
+
+	// Validate key format is a valid Kubernetes label key
+	keyPath := fldPath.Child("key")
+	allErrs = append(allErrs, metav1validation.ValidateLabelName(level.Key, keyPath)...)
+
+	return allErrs
+}
+
+// validateTopologyLevelOrder validates that topology levels are in the correct hierarchical order
+// (broadest to narrowest: Region > Zone > DataCenter > Block > Rack > Host > Numa).
+// This function assumes all domains have already been validated to be valid topology domains.
+// This validation also ensures domain uniqueness (duplicate domains would have the same order value).
+func validateTopologyLevelOrder(levels []grovecorev1alpha1.TopologyLevel, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for i := 1; i < len(levels); i++ {
+		prevLevel := levels[i-1]
+		currLevel := levels[i]
+
+		// Current level must be narrower than previous level
+		if currLevel.BroaderThan(prevLevel) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("domain"), currLevel.Domain,
+				fmt.Sprintf("topology levels must be in hierarchical order (broadest to narrowest). Domain '%s' at position %d cannot come after '%s' at position %d",
+					currLevel.Domain, i, prevLevel.Domain, i-1)))
+		} else if currLevel.Compare(prevLevel) == 0 {
+			allErrs = append(allErrs, field.Duplicate(fldPath.Index(i).Child("domain"), currLevel.Domain))
+		}
+	}
+
+	return allErrs
+}
+
+// validateTopologyLevelKeyUniqueness validates that all keys in the topology levels are unique.
+func validateTopologyLevelKeyUniqueness(levels []grovecorev1alpha1.TopologyLevel, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	seenKeys := make(map[string]int) // map key to first index where it appears
+
+	for i, level := range levels {
+		if firstIndex, exists := seenKeys[level.Key]; exists {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("key"), level.Key,
+				fmt.Sprintf("duplicate key: already used at levels[%d]", firstIndex)))
+		} else {
+			seenKeys[level.Key] = i
+		}
+	}
+
+	return allErrs
+}
+
+// ---------------------------- validate update of ClusterTopology -----------------------------------------------
+
+// validateUpdate validates the update to a ClusterTopology object.
+func (v *clusterTopologyValidator) validateUpdate(oldClusterTopology *grovecorev1alpha1.ClusterTopology) error {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, v.validateClusterTopologySpecUpdate(&v.clusterTopology.Spec, &oldClusterTopology.Spec, field.NewPath("spec"))...)
+	return allErrs.ToAggregate()
+}
+
+// validateClusterTopologySpecUpdate validates updates to the ClusterTopology specification.
+func (v *clusterTopologyValidator) validateClusterTopologySpecUpdate(newSpec, oldSpec *grovecorev1alpha1.ClusterTopologySpec, fldPath *field.Path) field.ErrorList {
```

Comment:
Sticking to the convention.

```suggestion
func (v *clusterTopologyValidator) validateClusterTopologySpecUpdate(oldSpec, newSpec *grovecorev1alpha1.ClusterTopologySpec, fldPath *field.Path) field.ErrorList {
```